Integrate ARIA element reflection in the Accessibility Tree

mrego opened the pull request:
#### 5a61ea9a825740c42f60af419813acb143df91f7
<pre>
Integrate ARIA element reflection in the Accessibility Tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=244972">https://bugs.webkit.org/show_bug.cgi?id=244972</a>

Reviewed by Ryosuke Niwa.

* LayoutTests/accessibility/element-reflection-ariaactivedescendant-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaactivedescendant.html: Added.
* LayoutTests/accessibility/element-reflection-ariacontrols-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariacontrols.html: Added.
* LayoutTests/accessibility/element-reflection-ariadescribedby-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariadescribedby.html: Added.
* LayoutTests/accessibility/element-reflection-ariadetails-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariadetails.html: Added.
* LayoutTests/accessibility/element-reflection-ariaerrormessage-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaerrormessage.html: Added.
* LayoutTests/accessibility/element-reflection-ariaflowto-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaflowto.html: Added.
* LayoutTests/accessibility/element-reflection-arialabelledby-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-arialabelledby.html: Added.
* LayoutTests/accessibility/element-reflection-ariaowns-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaowns.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations: Skip tests due to
  missing methods on WK1.
* LayoutTests/platform/win/TestExpectations: Skip new tests on Windows.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateRelationsIfNeeded): Take into account
 element reflection information.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore:: const): Ditto.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute): Expose method so
 it can be used from the a11y code.
(WebCore::Element::isElementsArrayReflectionAttribute): Ditto.
(WebCore::isElementReflectionAttribute): Deleted.
(WebCore::isElementsArrayReflectionAttribute): Deleted.
* Source/WebCore/dom/Element.h:

Canonical link: <a href="https://commits.webkit.org/254905@main">https://commits.webkit.org/254905@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c89fdaf64cbd509582947f58e0380177f65315a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35232 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/99967 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/158300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33729 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83009 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/96377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96306 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26841 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77477 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26683 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81620 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69709 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34822 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32634 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3427 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36400 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39349 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38321 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3733,6 +3733,15 @@ Vector<Element*> AccessibilityObject::elementsFromAttribute(const QualifiedName&
     if (!node || !node->isElementNode())
         return { };
 
+    if (Element* element = this->element()) {
```
auto* element

mrego replied:
Done.

fleizach commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3733,6 +3733,15 @@ Vector<Element*> AccessibilityObject::elementsFromAttribute(const QualifiedName&
         return { };
 
+    if (Element* element = this->element()) {
+        if (std::optional<Vector<RefPtr<Element>>> explicitlySetElements = element->getElementsArrayAttribute(attribute)) {
```
auto reflectedElements = element->

mrego replied:
Done.

fleizach commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3733,6 +3733,15 @@ Vector<Element*> AccessibilityObject::elementsFromAttribute(const QualifiedName&
+    if (Element* element = this->element()) {
+        if (std::optional<Vector<RefPtr<Element>>> explicitlySetElements = element->getElementsArrayAttribute(attribute)) {
+            Vector<Element*> elements;
+            for (RefPtr<Element> element : explicitlySetElements.value())
```
auto element

mrego replied:
Done.

fleizach commented on LayoutTests/accessibility/element-reflection.html:
```diff
@@ -0,0 +1,41 @@
+<!DOCTYPE html>
+<script src="../resources/js-test-pre.js"></script>
```
we no longer need the pre and post test js files

you can do

```
<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html>
<head>
<script src="../resources/accessibility-helper.js"></script>
<script src="../resources/js-test.js"></script>
</head>
<body>
```

mrego replied:
Done.

rniwa commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3733,6 +3733,15 @@ Vector<Element*> AccessibilityObject::elementsFromAttribute(const QualifiedName&
 
+    if (auto* element = this->element()) {
+        if (auto reflectedElements = element->getElementsArrayAttribute(attribute)) {
+            Vector<Element*> elements;
```
We can just use WTF::move instead.

rniwa commented on LayoutTests/accessibility/element-reflection.html:
```diff
@@ -0,0 +1,42 @@
+  }
+  customElements.define("x-input-idref", XInputIDREF);
+
+  input1.ariaLabelledByElements = [label1];
```
Please add tests for ariaActiveDescendantElement, ariaControlsElements, ariaDescribedByElements, ariaDetailsElements, ariaErrorMessageElement, ariaFlowToElements, and ariaOwnsElements.

mrego replied:
I've added tests for all of this but I have some some problems testing custom elements... I need to investigate if that's a WebKitGTK issue or also happens on Mac platforms. Uploaded new patch to check EWS though tests are still failing locally (running things manually on minibrowser shows the proper relationships on the a11y panel in the inspector).

mrego replied:
This issues are gone with #4561 so I'll update the test results using that patch, and this will have to land after that one.

rniwa commented on Source/WebCore/dom/Element.h:
```diff
@@ -131,6 +131,8 @@ class Element : public ContainerNode {
     WEBCORE_EXPORT void setElementAttribute(const QualifiedName& attributeName, Element* value);
     WEBCORE_EXPORT std::optional<Vector<RefPtr<Element>>> getElementsArrayAttribute(const QualifiedName& attributeName) const;
     WEBCORE_EXPORT void setElementsArrayAttribute(const QualifiedName& attributeName, std::optional<Vector<RefPtr<Element>>>&& value);
+    WEBCORE_EXPORT static bool isElementReflectionAttribute(const QualifiedName&);
```
Do we really need to export these two symbols???

mrego replied:
Yes, we don't need `WEBCORE_EXPORT` for these two. Removed it.

rniwa commented on LayoutTests/accessibility/element-reflection-ariacontrols.html:
```diff
@@ -0,0 +1,55 @@
+      tab.id = "innertab";
+      tab.role = "tab";
+      tab.textContent = "Tab 3";
+      tab.ariaControlsElements = [panel3];
```
This isn't really testing shadow DOM. We need to put the whole custom element into a shadow tree as I'm doing in https://github.com/WebKit/WebKit/pull/4561

mrego replied:
This is failing without this patch and fixed by this patch. It's referring an element outside the shadow tree from it.

I can put the whole thing in a shadow tree like in the your tests, but would that be something on top of this or a replacement? This is testing something that is currently broken, not sure what the new test will give us on top of this.

rniwa replied:
Interesting. If this test case is also failing then we need to add it as another test case.

mrego replied:
Sorry because my last comment here was misleading. "This is failing without this patch and fixed by this patch", as I mentioned on a previous comment I was having problems testing the Shadow DOM related parts of my tests, so this was not actually fixed by my patch.

The problems were happening on `run-webkit-tests` but not with `run-minibrowser` where I could see the relationships properly set on the a11y info in the inspector or using a screen reader with my patch (that's why I said "fixed by this patch").

The good thing is that those problems are fixed by #4561, this case and the one where the whole custom element is in the Shadow Tree would work the same. Not sure if we need 2 separated test cases or 1 is enough.

mrego replied:
I have added more test cases, some modifying  the `id` and still keeping the reference, and others moving an element out of a shadow tree and still keeping the reference working.

Please let me know if you think we need to test more cases, for now I'm marking this conversation as resolved.

rniwa commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4060,6 +4060,18 @@ void AXObjectCache::updateRelationsIfNeeded()
         // Collect all possible origins, i.e., elements with non-empty relation attributes.
         for (const auto& attribute : relationAttributes()) {
             auto& idsString = element.attributeWithoutSynchronization(attribute);
+            if (idsString.isEmpty()) {
```
This isn't right. Even when the id string isn't empty, we should be preferring the element reference over ID string.
Consider a case an element moves from being the same tree to being in an "outer" tree.

mrego replied:
Good point, even if we just simply change the `id` of the element that is being referenced, we'd be losing that relationship  from the a11y POV. So indeed we have to trust the element reference over the ID information. I'll change that and add tests.

Simple example that breaks due to this (we'll be getting null instead of the `foo/new-foo` element:
```html
<div id="target">target</div>
<div id="foo">foo</div>
<script>
  target.ariaDescribedByElements = [foo];
  foo.id = "new-foo";
  console.log(accessibilityController.accessibleElementById("target").ariaDescribedByElementAtIndex(0));
</script>
```

mrego replied:
And thinking about the `id` there might be another issue from the spec POV, what's going on if we have something like this?

```html
<div id="target">target</div>
<div id="foo1">foo1</div>
<div id="foo2">foo2</div>

<script>
  let myFoo1 = document.getElementById("foo1");
  let myFoo2 = document.getElementById("foo2");
  
  target.ariaDescribedByElements = [myFoo1];
  
  myFoo1.id = "foo2";
  myFoo2.id = "foo1";
  
  console.log("aria-describedby attribute: " + target.getAttribute("aria-describedby"));
  console.log("ariaDescribedByElements: " + target.ariaDescribedByElements[0].id);
</script>
```

Should we trust the attribute or the element reference? I guess I need to fill another HTML issue about this. :/

mrego replied:
JFTR filled https://github.com/whatwg/html/issues/8307

mrego replied:
I've changed this on the implementation so it follows element reference and not the ID. That's also what Chromium does.

rniwa commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -3739,9 +3739,24 @@ Vector<Element*> AccessibilityObject::elementsFromAttribute(const QualifiedName&
+    } else if (Element::isElementsArrayReflectionAttribute(attribute)) {
+        if (auto reflectedElements = element.getElementsArrayAttribute(attribute)) {
+            Vector<Element*> elements;
+            for (auto reflectedElement : reflectedElements.value())
```
Use WTF::map?

mrego replied:
Good idea. Used it also a few lines below too.

rniwa reviewed (approved):
r=me

mrego committed:
Integrate ARIA element reflection in the Accessibility Tree
https://bugs.webkit.org/show_bug.cgi?id=244972

Reviewed by Ryosuke Niwa.

* LayoutTests/accessibility/element-reflection-ariaactivedescendant-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaactivedescendant.html: Added.
* LayoutTests/accessibility/element-reflection-ariacontrols-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariacontrols.html: Added.
* LayoutTests/accessibility/element-reflection-ariadescribedby-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariadescribedby.html: Added.
* LayoutTests/accessibility/element-reflection-ariadetails-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariadetails.html: Added.
* LayoutTests/accessibility/element-reflection-ariaerrormessage-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaerrormessage.html: Added.
* LayoutTests/accessibility/element-reflection-ariaflowto-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaflowto.html: Added.
* LayoutTests/accessibility/element-reflection-arialabelledby-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-arialabelledby.html: Added.
* LayoutTests/accessibility/element-reflection-ariaowns-expected.txt: Added.
* LayoutTests/accessibility/element-reflection-ariaowns.html: Added.
* LayoutTests/platform/mac-wk1/TestExpectations: Skip tests due to
  missing methods on WK1.
* LayoutTests/platform/win/TestExpectations: Skip new tests on Windows.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateRelationsIfNeeded): Take into account
 element reflection information.
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore:: const): Ditto.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isElementReflectionAttribute): Expose method so
 it can be used from the a11y code.
(WebCore::Element::isElementsArrayReflectionAttribute): Ditto.
(WebCore::isElementReflectionAttribute): Deleted.
(WebCore::isElementsArrayReflectionAttribute): Deleted.
* Source/WebCore/dom/Element.h:

Canonical link: https://commits.webkit.org/254905@main
